dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/FaultToleranceCDIComponent.java

```diff
@@ -63,26 +67,32 @@ protected void deactivate() {
     }
 
     /**
-     * Get the current metric provider
+     * Get the current metric providers
      * <p>
      * The provider may change if features are updated
+     * 
+     * @return an unmodifiable list of metric providers. If any provider is not a 
+     *         dummy provider then all dummy providers will be filtered out
      */
-    public static MetricRecorderProvider getMetricProvider() {
-        return metricProvider;
+    public static List<MetricRecorderProvider> getMetricProviders() {
+        
+        if (metricProviders.stream().anyMatch(mp -> ! mp.isDummy())) {
+            return Collections.unmodifiableList(metricProviders.stream()
+                                                .filter(mp -> ! mp.isDummy())
+                                                .collect(Collectors.toList()));
+        }
+        
+        return Collections.unmodifiableList(metricProviders);
     }
```

Comment:
I don't like this `isDummy()` flag - I think it makes the code unnecessarily more complex. You've got several things provided via service, but you're picking out one to be special. Using a service like this generally only makes sense if you can treat all instances the same (which was true in the original code, but isn't true now).

Also, it doesn't work right now because `DummyRecorderProvider` hasn't been updated to override `isDummy()`

I think it would be better to:
- remove `isDummy()`
- remove `DummyRecorderProvider` so that it's not provided via a service
  - you do still need a `MetricRecorderProvider` that returns the `DummyMetricRecorder`, but it can be a singleton rather than a component since we don't need it to be registered as a service.
  - I would probably make `DummyMetricRecorder` implement `MetricRecorderProvider` as well and return itself.
- if there are no `MetricRecorderProviders`, return a provider for the `DummyMetricRecorder` instead

Reply:
Hmm, you do also need to be careful of the `metricProviders` list changing while you're reading it. It's a CopyOnWrite list, but it can still change between one call and the next.

It might be easier to return the list and do the check of "do I have any metric recorders" in `AggregatedFTPolicy` instead, after iterating through the providers. That would also mean you don't need a provider for the `DummyMetricRecorder`.

Reply:
Actually you can just not call `setMetricRecorder` if there are no recorders. `AbstractExecutorBuilderImpl` already defaults to `DummyMetricRecorder` if that method is not called.

Reply:
> Actually you can just not call setMetricRecorder if there are no recorders.

This _will_ make things simpler.